File: Data Structures & Algorithms/binary-search/submission-1.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int p1 = 0;
        int p2 = nums.size() - 1;
        while (true){
            int middleIndex = p1 + ((p2 - p1) / 2);
            int val = nums[middleIndex];
            if (nums[p1] == target){
                return p1;
            }else if (nums[p2] == target){
                return p2;
            }
            if (middleIndex == p1) {
                return -1;
            }
            if (val == target) {
                return middleIndex;
            } else if (val > target) {
                p2 = middleIndex;
            } else {
                p1 = middleIndex;
            }
        }

        return -1;
    }
};

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l,r = 0, len(nums) - 1
        while l <= r:
            mid = l + (r - l) // 2
            if nums[l] == target:
                return l
            elif nums[r] == target:
                return r
            elif nums[mid] == target:
                return mid
            
            if nums[l] < nums[mid]:
                if nums[l] < target < nums[mid]:
                    r = mid - 1
                    l += 1
                else:
                    l = mid + 1
                    r -= 1
            else:
                if nums[mid] < target < nums[r]:
                    l = mid + 1
                    r -= 1
                else:
                    r = mid - 1
                    l += 1
        return -1